File: java/UpDownBot.java
import com.google.gson.JsonObject;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class UpDownBot {

    public static void main(String[] args) {
        upDownBot();
    }

    public static void upDownBot() {
        Api.Simulation simulation = new Api.Simulation(
                "secondspace2025",
                "tiny_random",
                "updown-java-bot",
                "bob@mail.com",
                false,
                "https://dianti.de.r.appspot.com/api"
        );
        JsonObject currentState = simulation.initialState;
        Map<String, Boolean> directions = new HashMap<>();

        while (currentState.get("running").getAsBoolean()) {
            List<Api.Command> commands = new ArrayList<>();
            JsonArray elevators = currentState.getAsJsonArray("elevators");
            for (int i = 0; i < elevators.size(); i++) {
                JsonObject elevator = elevators.get(i).getAsJsonObject();
                String elevatorId = elevator.get("id").getAsString();
                boolean direction = directions.getOrDefault(elevatorId, Api.UP);

                if (direction == Api.UP && elevator.get("floor").getAsInt() == simulation.numFloors) {
                    direction = Api.DOWN;
                } else if (direction == Api.DOWN && elevator.get("floor").getAsInt() == 1) {
                    direction = Api.UP;
                }
                directions.put(elevatorId, direction);

                boolean action = Api.MOVE;
                JsonArray buttonsPressed = elevator.getAsJsonArray("buttons_pressed");
                if (buttonsPressed.contains(new JsonPrimitive(elevator.get("floor").getAsInt()))) {
                    action = Api.STOP;
                } else {
                    JsonArray requests = currentState.getAsJsonArray("requests");
                    for (int j = 0; j < requests.size(); j++) {
                        JsonObject request = requests.get(j).getAsJsonObject();
                        if (request.get("floor").getAsInt() == elevator.get("floor").getAsInt() &&
                                request.get("direction").getAsBoolean() == direction) {
                            action = Api.STOP;
                            break;
                        }
                    }
                }
                commands.add(new Api.Command(elevatorId, direction, action));
            }
            currentState = simulation.send(commands);
        }

        System.out.println("Score: " + currentState.get("score").getAsInt());
        System.out.println("Replay URL: " + currentState.get("replay_url").getAsString());
    }
}

File: c++/updown_bot.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include "api.hpp"

void updown_bot() {
    Simulation sim("secondspace2025", "tiny_random", "updown-cpp-bot", "bob@mail.com", true);
    json current_state = sim.get_initial_state();
    std::map<std::string, bool> directions;

    while (current_state["running"]) {
        std::vector<Command> commands;
        for (const auto& elevator : current_state["elevators"]) {
            std::string elevator_id = elevator["id"];
            bool direction = directions[elevator_id];
            if (direction == UP && elevator["floor"] == sim.get_num_floors()) {
                direction = DOWN;
            } else if (direction == DOWN && elevator["floor"] == 1) {
                direction = UP;
            }
            directions[elevator_id] = direction;

            bool action = MOVE;
            for (const auto& floor : elevator["buttons_pressed"]) {
                if (floor == elevator["floor"]) {
                    action = STOP;
                    break;
                }
            }
            if (action == MOVE) {
                for (const auto& request : current_state["requests"]) {
                    if (request["floor"] == elevator["floor"] && request["direction"] == direction) {
                        action = STOP;
                        break;
                    }
                }
            }

            commands.push_back({elevator_id, direction, action});
        }
        current_state = sim.send(commands);
    }

    std::cout << "Score: " << current_state["score"] << std::endl;
    std::cout << "Replay URL: " << current_state["replay_url"] << std::endl;
}

int main() {
    updown_bot();
    return 0;
}
